src/libraries/System.Private.CoreLib/src/System/Globalization/TextInfo.cs

```diff
@@ -566,17 +588,16 @@ public string ToUpper(string str)
             return ChangeCaseCommon<ToUpperConversion>(str);
         }
 
-        internal static char ToUpperAsciiInvariant(char c)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static char ToUpperAsciiInvariant(char c)
         {
-            if ((uint)(c - 'a') <= (uint)('z' - 'a'))
+            if (UnicodeUtility.IsInRangeInclusive(c, 'a', 'z'))
             {
-                c = (char)(c & ~0x20);
+                c = (char)(c & 0x5F); // = low 7 bits of ~0x20
```

Comment:
This saves a mov?

Reply:
It's not the mov _per se_ (since those can be free). It's the extension of 16-bit to 32-bit, which the JIT implicitly does before the return. Those are sometimes weirdly expensive on x86 and can stall pipelining.

This trick allows the JIT to realize that the return value already fits neatly within 16 bits, so it can skip the extension step and return the 32-bit register as is.